llvm/lib/Transforms/Utils/ScalarEvolutionExpander.cpp

```diff
@@ -161,6 +161,11 @@ SCEVExpander::findInsertPointAfter(Instruction *I,
   return IP;
 }
 
+bool SCEVExpander::expansionMayIntroduceUB(const SCEV *Expr) {
+  return SCEVExprContains(Expr,
+                          [](const SCEV *Op) { return isa<SCEVUDivExpr>(Op); });
```

Comment:
Is this the same as SafeToHoist in SCEVExpander::expand?

We could probably do something more clever if it mattered: the only way for the exit count to contain UB is if the branch never executes, so it would be correct to just transform `a/b` to `a/umax(b, 1)`.

Reply:
This is `SCEVExpander::isSafeToExpand()`. (I think the SafeToHoist stuff is some half-dead leftover to handle call-sites that fail to call isSafeToExpand.)

Reply:
Updated the PR to also check if the divisor is guaranteed to be non-zero. This may miss some safe cases, e.g. where the division happens outside the loop already. Not sure if we have any good way of checking that, except possibly the checks if there are existing equivalent instructions outside the loop during expansion?

> We could probably do something more clever if it mattered: the only way for the exit count to contain UB is if the branch never executes, so it would be correct to just transform a/b to a/umax(b, 1).

I think in some cases this may not be correct, e.g. we have two exit counts 1) c and 2) `a - (42/b)`, with the BTC being `umin(c, a - (42/b))`. If `b == 0` then the BTC should be `c`, but using 1 as divisor may result in `a - (42/(umax 1, b))` < `c` and we  would use an incorrect BTC?

Reply:
> e.g. we have two exit counts 1) `c` and 2) `a - (42/b)`, with the BTC being `umin(c, a - (42/b))`. If `b == 0` then the BTC should be `c`, but using 1 as divisor may result in `a - (42/(umax 1, b)) < c` and we would use an incorrect BTC?

In the given example, `c` has to be zero, I think?  If `c` isn't zero, and `b` is zero, we divide by zero even before any optimizations run.

Reply:
> Updated the PR to also check if the divisor is guaranteed to be non-zero. 

Can we not use isSafeToExpand() for some reason?

Reply:
Sorry for just coming back to this now!

> In the given example, c has to be zero, I think? If c isn't zero, and b is zero, we divide by zero even before any optimizations run.

Ah yes, updated to rewrite the expression, thanks!

> Can we not use isSafeToExpand() for some reason?

Updated as well in the latest version, thanks!

Reply:
There's one more case here I wasn't thinking about: udiv where the RHS is poison.  umax isn't sufficient in that case because umax(1,poison) is still poison,